docs/graph/concept.md

```diff
@@ -0,0 +1,302 @@
+
+若 $G$ 与 $H$ 的点集本身不相交，则 $G \cup H = G + H$ 。
+
+比如，森林可以定义成若干棵树的和。
+
+???+note "并与和的区别"
+    可以理解为，“并”会让两张图中“名字相同”的点、边合并，而“和”则不会。
+
+## 特殊的点集/边集
+
+### 支配集
+
+对于无向图 $G=(V, E)$ ，若 $V'\subseteq V$ 且 $\forall v\in(V\setminus V')$ 存在边 $(u, v)\in E$ 满足 $u\in V'$ ，则 $V'$ 是图 $G$ 的一个 **支配集 (Dominating set)** 。
+
+无向图 $G$ 最小的支配集的大小记作 $\gamma(G)$ 。求一张图的最小支配集是 NP 困难的。
+
+??? note "什么是 NP 困难？"
+    如果所有 NP 问题都可以在多项式时间内规约到问题 $H$ ，则问题 $H$ 是 NP 困难的。需要注意的是，NPC 问题必须是判定问题，因此求一张图的最小支配集 **不是** NPC 问题。实际上，NPC 问题是 NP 问题和 NP 困难问题的交，而 NP 问题必须是判定问题。NP 困难的最优化问题往往有一个对应的判定版本是 NPC 问题，如最小支配集问题的判定版本是给定一张图和一个数 $K$ ，判定是否存在一个大小不超过 $K$ 的支配集，这个问题是一个 NPC 问题。
+
+    详见 [计算理论基础](../misc/cc-basic.md) 。
+
+对于有向图 $G=(V, E)$ ，若 $V'\subseteq V$ 且 $\forall v\in(V\setminus V')$ 存在边 $(u, v)\in E$ 满足 $u\in V'$ ，则 $V'$ 是图 $G$ 的一个 **出 - 支配集 (Out-dominating set)** 。类似地，可以定义有向图的 **入 - 支配集 (In-dominating set)** 。
+
+有向图 $G$ 最小的出 - 支配集大小记作 $\gamma^+(G)$ ，最小的入 - 支配集大小记作 $\gamma^-(G)$ 。
+
+### 边支配集
+
+对于图 $G=(V, E)$ ，若 $E'\subseteq E$ 且 $\forall e\in(E\setminus E')$ 存在 $E'$ 中的边与其有公共点，则称 $E'$ 是图 $G$ 的一个 **边支配集 (Edge dominating set)** 。
+
+求一张图的最小边支配集是 NP 困难的。
+
+### 独立集
+
+对于图 $G=(V, E)$ ，若 $V'\subseteq V$ 且 $V'$ 中任意两点都不相邻，则 $V'$ 是图 $G$ 的一个 **独立集 (Independent set)** 。
+
+图 $G$ 最大的独立集的大小记作 $\alpha(G)$ 。求一张图的最大独立集是 NP 困难的。
+
+### 匹配
+
+对于图 $G=(V, E)$ ，若 $E'\in E$ 且 $E'$ 中任意两条不同的边都没有公共的端点，且 $E'$ 中任意一条边都不是自环，则 $E'$ 是图 $G$ 的一个 **匹配 (Matching)** ，也可以叫作 **边独立集 (Independent edge set)** 。如果一个点是匹配中某条边的一个端点，则称这个点是 **被匹配的 (matched)/饱和的 (saturated)** ，否则称这个点是 **不被匹配的 (unmatched)** 。
+
+边数最多的匹配被称作一张图的 **最大匹配 (Maximum-cardinality matching)** 。图 $G$ 的最大匹配的大小记作 $\nu(G)$ 。
+
+如果边带权，那么权重之和最大的匹配被称作一张图的 **最大权匹配 (Maximum-weight matching)** 。
+
+如果一个匹配在加入任何一条边后都不再是一个匹配，那么这个匹配是一个 **极大匹配 (Maximal matching)** 。最大的极大匹配就是最大匹配，任何最大匹配都是极大匹配。极大匹配一定是边支配集，但边支配集不一定是匹配。最小极大匹配和最小边支配集大小相等，但最小边支配集不一定是匹配。求最小极大匹配是 NP 困难的。
+
+如果在一个匹配中所有点都是被匹配的，那么这个匹配是一个 **完美匹配 (Perfect matching)** 。如果在一个匹配中只有一个点不被匹配，那么这个匹配是一个 **准完美匹配 (Near-perfect matching)** 。
+
+求一张普通图或二分图的匹配或完美匹配个数都是 #P 完全的。
+
+对于一个匹配 $M$ ，若一条路径以非匹配点为起点，每相邻两条边的其中一条在匹配中而另一条不在匹配中，则这条路径被称作一条 **交替路径 (Alternating path)** ；一条在非匹配点终止的交替路径，被称作一条 **增广路径 (Augmenting path)** 。
+
+### 点覆盖
+
+对于图 $G=(V, E)$ ，若 $V'\subseteq V$ 且 $\forall e\in E$ 满足 $e$ 的至少一个端点在 $V'$ 中，则称 $V'$ 是图 $G$ 的一个 **点覆盖 (Vertex cover)** 。
+
+一个点集是点覆盖的充要条件是其补集是独立集，因此最小点覆盖的补集是最大独立集。求一张图的最小点覆盖是 NP 困难的。
+
+### 边覆盖
+
+对于图 $G=(V, E)$ ，若 $E'\subseteq E$ 且 $\forall v\in V$ 满足 $v$ 与 $E'$ 中的至少一条边相邻，则称 $E'$ 是图 $G$ 的一个 **边覆盖 (Edge cover)** 。
+
+最小边覆盖的大小记作 $\rho(G)$ ，可以由最大匹配贪心扩展求得。特别地，完美匹配一定是一个最小边覆盖。对于图 $G$ ，其最大匹配的大小不超过最小边覆盖的大小，即 $\nu(G)\le\rho(G)$ 。
+
+### 团
+
+对于图 $G=(V, E)$ ，若 $V'\subseteq V$ 且 $V'$ 中任意两个不同的顶点都相邻，则 $V'$ 是图 $G$ 的一个 **团 (Clique)** 。团的导出子图是完全图。
+
+如果一个团在加入任何一个顶点后都不再是一个团，则这个团是一个 **极大团 (Maximal clique)** 。
+
+一张图的最大团的大小记作 $\omega(G)$ ，最大团的大小等于其补图最大独立集的大小，即 $\omega(G)=\alpha(\bar{G})$ 。求一张图的最大团是 NP 困难的。
```

Comment:
同样的，这里也少一个结论 ` $V^*$ 为 $G$ 的团当且仅当 $V^*$ 为 $\bar{G}$ 的独立集。`。
这个 note 也不全，不知道是我没找到还是就没有放进这个 pr

![image](https://user-images.githubusercontent.com/10709657/69007811-c4784600-097d-11ea-8109-5eed68643c84.png)

Reply:
极大团，也可以求补图的极大独立集。前面的极小就是最小吧..

> A single maximal clique can be found by a straightforward greedy algorithm. Starting with an arbitrary clique (for instance, any single vertex or even the empty set), grow the current clique one vertex at a time by looping through the graph's remaining vertices. For each vertex v that this loop examines, add v to the clique if it is adjacent to every vertex that is already in the clique, and discard v otherwise. This algorithm runs in linear time.